core/common/src/main/java/alluxio/master/PollingMasterInquireClient.java

```diff
@@ -129,31 +143,81 @@ private InetSocketAddress getAddress() {
       addresses = mConnectDetails.getAddresses();
     }
 
+    if (mConfiguration.getBoolean(PropertyKey.USER_MASTER_POLLING_CONCURRENT)) {
+      return findActiveAddressConcurrent(addresses);
+    } else {
+      return findActiveAddress(addresses);
+    }
+  }
+
+  @Nullable
+  private InetSocketAddress findActiveAddressConcurrent(List<InetSocketAddress> addresses) {
+    List<Future<InetSocketAddress>> futures = new ArrayList<>(addresses.size());
+    try {
+      ExecutorCompletionService<InetSocketAddress> completionService =
+          new ExecutorCompletionService<>(EXECUTOR_SERVICE);
+      for (InetSocketAddress address : addresses) {
```

Comment:
though we run those three polling tasks concurrently here essentially we have to wait all of them to finish. I'm thinking if we can do better and return it immediately as soon as a validate result comes up. 

I'm thinking if we can leverage the completable future. Here are some examples:
https://stackoverflow.com/questions/36612224/in-java-how-do-i-process-completablefutures-and-get-the-first-desireable-result

Reply:
Please see line 164. The ExecutorCompletionService#take method retrieves the first Future that has completed and, if its value is not null, returns it directly while canceling any remaining Futures at line 175.

Reply:
ah okay sounds good